File: languages/ts/src/terminal_progress.ts
const ANSI_ESC: string = "\x1B"
const ANSI_OSC: string = ANSI_ESC + "]"

const ANSI_ST: string = ANSI_ESC + "\\"
const ANSI_BELL: string = "\x07"

const OSC_PROGRESS_REPORT_BASE: string = "9;4"

function toStringCapped(num: number): string {
	if (num < 0) {
		return (0).toFixed(0)
	}

	if (num > 100) {
		return (100).toFixed(0)
	}

	return num.toFixed(0)
}

function send_progress_impl(st: U8, pr: Optional<U8> = undefined): void {
	if (pr === undefined || pr === null) {
		process.stdout.write(
			ANSI_OSC +
				OSC_PROGRESS_REPORT_BASE +
				";" +
				st.toString() +
				ANSI_BELL
		)
	} else {
		process.stdout.write(
			ANSI_OSC +
				OSC_PROGRESS_REPORT_BASE +
				";" +
				st.toString() +
				";" +
				toStringCapped(pr) +
				ANSI_BELL
		)
	}
}

export type Optional<T> = T | undefined | null

export enum ProgressState {
	ProgressStateRemove = 0,
	ProgressStateSet = 1,
	ProgressStateError = 2,
	ProgressStateIndeterminate = 3,
	ProgressStatePaused = 4,
}

export type U8 = number

export interface ProgressReportRemove {
	type: ProgressState.ProgressStateRemove
}

export interface ProgressReportSet {
	type: ProgressState.ProgressStateSet
	value: U8
}

export interface ProgressReportError {
	type: ProgressState.ProgressStateError
	value: Optional<U8>
}

export interface ProgressReportIndeterminate {
	type: ProgressState.ProgressStateIndeterminate
}

export interface ProgressReportPaused {
	type: ProgressState.ProgressStatePaused
	value: Optional<U8>
}

export type ProgressReport =
	| ProgressReportRemove
	| ProgressReportSet
	| ProgressReportError
	| ProgressReportIndeterminate
	| ProgressReportPaused

export function send_progress(report: ProgressReport): void {
	if (!process.stdout.isTTY) {
		return
	}

	switch (report.type) {
		case ProgressState.ProgressStateRemove: {
			send_progress_impl(report.type)
			break
		}
		case ProgressState.ProgressStateSet: {
			send_progress_impl(report.type, report.value)
			break
		}
		case ProgressState.ProgressStateError: {
			send_progress_impl(report.type, report.value)
			break
		}
		case ProgressState.ProgressStateIndeterminate: {
			send_progress_impl(report.type)
			break
		}
		case ProgressState.ProgressStatePaused: {
			send_progress_impl(report.type, report.value)
			break
		}
		default: {
			return
		}
	}
}

File: languages/cpp/src/terminal_progress.cpp


#include "terminal_progress_cpp.hpp"

#include <fstream>
#include <iostream>

#define ANSI_ESC "\x1B"
#define ANSI_OSC ANSI_ESC "]"

#define ANSI_ST ANSI_ESC "\\"
#define ANSI_BELL "\x07"

#define OSC_PROGRESS_REPORT_BASE "9;4"

static void send_progress_impl(std::ostream &os, uint8_t st,
                               std::optional<uint8_t> pr) {

  if (pr.has_value()) {
    os << ANSI_OSC OSC_PROGRESS_REPORT_BASE << ";" << st << ";" << pr.value()
    << ANSI_BELL;
    
  } else {
    os << ANSI_OSC OSC_PROGRESS_REPORT_BASE << ";" << st << ANSI_BELL;
  }

  std::flush(os);
}

static bool is_ostream_atty(std::ostream &os) {

  const auto fos = reinterpret_cast<std::ofstream *>(&os);

  if (fos == nullptr) {
    return false;
  }

#if __cplusplus >= 202600L && defined(__cpp_lib_fstream_native_handle)
  const auto fd = fos->native_handle();
  return isatty(fd);
#else
  return true;
#endif
}

void ProgressReport::send_to(std::ostream &os) const {

  if (!is_ostream_atty(os)) {
    return;
  }

  switch (this->m_type) {
  case ProgressStateRemove: {
    send_progress_impl(os, this->m_type, std::nullopt);
    break;
  }
  case ProgressStateSet: {
    send_progress_impl(os, this->m_type, this->m_data.set);
    break;
  }
  case ProgressStateError: {
    send_progress_impl(os, this->m_type, this->m_data.error);
    break;
  }
  case ProgressStateIndeterminate: {
    send_progress_impl(os, this->m_type, std::nullopt);
    break;
  }
  case ProgressStatePaused: {
    send_progress_impl(os, this->m_type, this->m_data.paused);
    break;
  }
  default: {
    return;
  }
  }
}

void ProgressReport::send() const { return this->send_to(std::cout); }

std::ostream &operator<<(std::ostream &os, ProgressReport report) {
  report.send_to(os);
  return os;
}

void send_progress_to_cpp(ProgressReport report, std::ostream &os) {
  return report.send_to(os);
}

void send_progress_cpp(ProgressReport report) { return report.send(); }

ProgressReport::ProgressReport(ProgressState state) : m_type{state}, m_data{} {}

ProgressReport::ProgressReport(ProgressState state, details::DataUnion data)
    : m_type{state}, m_data{data} {}

[[nodiscard]] ProgressReport ProgressReport::remove() {
  return ProgressReport{ProgressState::ProgressStateRemove};
}

[[nodiscard]] ProgressReport ProgressReport::set(std::uint8_t value) {
  return ProgressReport{ProgressState::ProgressStateSet,
                        details::DataUnion{.set = value}};
}

[[nodiscard]] ProgressReport ProgressReport::error() {
  return ProgressReport::error(std::nullopt);
}

[[nodiscard]] ProgressReport ProgressReport::error(uint8_t value) {
  return ProgressReport::error(std::optional<std::uint8_t>{value});
}

[[nodiscard]] ProgressReport
ProgressReport::error(std::optional<std::uint8_t> value) {
  return ProgressReport{ProgressState::ProgressStateError,
                        details::DataUnion{.error = value}};
}

[[nodiscard]] ProgressReport ProgressReport::indeterminate() {
  return ProgressReport{ProgressState::ProgressStateIndeterminate};
}

[[nodiscard]] ProgressReport ProgressReport::paused() {
  return ProgressReport::paused(std::nullopt);
}

[[nodiscard]] ProgressReport ProgressReport::paused(uint8_t value) {
  return ProgressReport::paused(std::optional<std::uint8_t>{value});
}

[[nodiscard]] ProgressReport
ProgressReport::paused(std::optional<std::uint8_t> value) {
  return ProgressReport{ProgressState::ProgressStatePaused,
                        details::DataUnion{.paused = value}};
}
